File: Java/Stack.java
/**
 ** Stack implementation with stack size variable
 ** Singly linked list ; generic stack type
 ** Mohammed Bataineh
 **/
public class Stack<T> {

  // node for singly linked list
  private class Node {
    T item;
    Node prev;
  }

  // class variables
  private Node top;
  private int size;

  // return size
  public int getSize() {
    return size;
  }

  // First node in list must point to null
  // Initialize top as null is required for first push
  Stack() {
    top = null;
  }

  // add a node onto list
  public void push(T item) {
    Node current = new Node();
    current.item = item;
    current.prev = top;
    top = current;
    size++;
  }

  // remove top node
  public void pop() {
    // if size is 0
    if(size == 0){
      System.out.println("Stack is empty.");
      return;
    }
    top = top.prev;
    size--;
  }

  // Alternative pop method:
  // this one allows for pop to return the item as well
  // Disadvantage: if T is a large object then the copy
  // may result in performance reductions
  /*
  public T pop() {
    T item = top;
    top = top.prev;
    return item;
    }
  */

  // check top item
  public T peek() {
    // if size is 0
    if(size == 0){
      System.out.println("Stack is empty.");
      return null;
    }
    return top.item;
  }

  // checks if size is 0
  public boolean isEmpty(){
    if(size == 0){
      return true;
    }
    else{
      return false;
    }
  }

}

File: C/Stack.c
//#include <limits.h>
#include <stdio.h>
#include <stdlib.h>

// what kind of stack do you want it to be?
// define type of "item"
#define TYPE int

// Tracks size of stack
int stackSize = 0;

// Simpl node structure for stack
struct StackNode {
  TYPE item;
  struct StackNode* prev;
};

// method to create new node
struct StackNode* createNode(TYPE item) {
  struct StackNode* node = (struct StackNode*)malloc(sizeof(struct StackNode));
  node->item = item;
  node->prev = NULL;
  return node;
}

// Checks if empty
int isEmpty(struct StackNode* top) {
  // if top is NULL then it is not false (true)
  return !top;
}

// Adds node onto stack
void push(struct StackNode** top, TYPE item){
  struct StackNode* stackNode = createNode(item);
  stackNode->prev = *top;
  *top = stackNode;
  stackSize++;
}

// Pops node off of stack and returns item value
TYPE pop(struct StackNode** top) {
  if (isEmpty(*top)){
    fprintf(stderr, "Stack is empty; Cannot peek.");
    return NULL;
  }
  struct StackNode* temp = *top;
  *top = (*top)->prev;
  TYPE popped = temp->item;
  free(temp);
  stackSize--;
  return popped;
}

// returns top node item
TYPE peek(struct StackNode* top) {
  if (isEmpty(top)) {
    fprintf(stderr, "Stack is empty; Cannot peek.");
    return NULL;
  }
  return (top)->item;
}

// Main
int main()
{
    struct StackNode* top = NULL;

    push(&top, 10);
    push(&top, 20);
    push(&top, 30);

    printf("%d popped from stack\n", pop(&top));

    printf("Top element is %d\n", peek(top));

    printf("Size is %d\n", stackSize);

    return 0;
}
